langserver/fs.go

```diff
@@ -124,14 +124,24 @@ func (h *overlay) didChange(params *lsp.DidChangeTextDocumentParams) error {
 		return fmt.Errorf("received textDocument/didChange for unknown file %q", params.TextDocument.URI)
 	}
 
-	for _, change := range params.ContentChanges {
+	contents, err := ApplyContentChanges(params.TextDocument.URI, contents, params.ContentChanges)
+	if err != nil {
+		return err
+	}
+
+	h.set(params.TextDocument.URI, contents)
+	return nil
+}
+
+// ApplyContentChanges updates `contents` based on `changes`
+func ApplyContentChanges(uri lsp.DocumentURI, contents []byte, changes []lsp.TextDocumentContentChangeEvent) ([]byte, error) {
```

Comment:
Can you not export this function? I don't think we have a need to use it outside of this pkg yet.